context/context.go

```diff
@@ -0,0 +1,76 @@
+package context
+
+import (
+	gocontext "context"
+	"fmt"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Context interface {
+	gocontext.Context
+	WithGoContext(c gocontext.Context) Context
+	Errorf(err error)
+	StartSpan(tracer trace.Tracer, spanName string) (Context, trace.Span)
+	SetSpanAttributes(key string, val any)
+}
+
+func NewContext(ctx gocontext.Context) Context {
+	return &context{
+		Context: ctx,
+	}
+}
+
+// context implements Context
+type context struct {
+	gocontext.Context
+}
+
+func (c context) WithGoContext(cx gocontext.Context) Context {
+	c.Context = cx
+	return &c
+}
+
+func (c context) StartSpan(tracer trace.Tracer, spanName string) (Context, trace.Span) {
+	traceCtx, span := tracer.Start(c.Context, spanName)
+	c.Context = traceCtx
+	return &c, span
+}
+
+func (c *context) SetSpanAttributes(key string, val any) {
+	var attr attribute.Value
+	switch v := val.(type) {
+	case bool:
+		attr = attribute.BoolValue(v)
+	case string:
+		attr = attribute.StringValue(v)
+	case int:
+		attr = attribute.IntValue(v)
+	case int32:
+		attr = attribute.IntValue(int(v))
+	case int64:
+		attr = attribute.Int64Value(v)
+	case float64:
+		attr = attribute.Float64Value(v)
+	case []bool:
+		attr = attribute.BoolSliceValue(v)
+	case []int64:
+		attr = attribute.Int64SliceValue(v)
+	case []float64:
+		attr = attribute.Float64SliceValue(v)
+	case []string:
+		attr = attribute.StringSliceValue(v)
+	default:
+		attr = attribute.StringValue(fmt.Sprintf("%v", v))
+	}
+
+	trace.SpanFromContext(c).SetAttributes(attribute.KeyValue{Key: attribute.Key(key), Value: attr})
+}
+
+func (c *context) Errorf(err error) {
+	span := trace.SpanFromContext(c)
+	span.RecordError(err)
+	span.SetStatus(codes.Error, err.Error())
```

Comment:
I think we should also logger.Errorf here

Reply:
I think we shouldn't. I wouldn't want to log in all the layers. On the other hand, tracing should be done on all the layers.